### PR TITLE
Stop creating a conf.py automatically and doing magic around README handling

### LINKED ISSUE #2483: Stop creating a conf.py/mkdocs.yaml file automatically for users and doing magic
We should stop adding a `conf.py` file for projects that don't have one.  We also used to do magic around README handling, we should remove it:

https://github.com/rtfd/readthedocs.org/blob/535f31540fea36a321d148cfee302b34e90adfc9/readthedocs/doc_builder/base.py#L97-L116

### LINKED ISSUE #10627: Build: fail build if conf.py file doesn't exist
## Details
I am trying to build the docs for the first time and I am receiving the following error:

`root file /home/docs/checkouts/readthedocs.org/user_builds/gestrec/checkouts/latest/docs/index.rst not found
`
The docs build locally without any warnings.

* Read the Docs project URL: https://readthedocs.org/projects/gestrec/
* Build URL (if applicable): [https://readthedocs.org/projects/gestrec/builds/21591531/](https://readthedocs.org/projects/gestrec/builds/21591531/)
* Read the Docs username (if applicable): https://readthedocs.org/profiles/witolddebski/

## Expected Result

docs build successfully, just like locally

## Actual Result

I got the above error. I noticed that the filepath in the error is "_.../latest/docs/index.rst_" and all my source files are in folder _docs/source_, so I set the following variables to try and fix the issue:
```
master_doc = 'index'
root_doc = 'index'
```
I also tried:
```
master_doc = 'source/index'
root_doc = 'source/index'
```
it did not fix the issue.

I suspect it is somehow not picking up my conf.py file, since I am also getting an error `/home/docs/checkouts/readthedocs.org/user_builds/gestrec/checkouts/latest/docs/source/reference.rst:4: ERROR: Unknown directive type "automodule".` which I'm not getting locally, as I have the extension defined in _conf.py_

help plz

### PR DESCRIPTION
**Pull request by saadmk11**
I need some input about the README handling. if that's what we want or not?

* Closes #2483 
* Closes #10627

### CODE REVIEW

File: readthedocs/doc_builder/backends/sphinx.py [hunk 1]
```diff
@@ -56,20 +56,6 @@ def __init__(self, *args, **kwargs):
...
-        conf_template = render_to_string(
-            'sphinx/conf.py.conf',
-            {
-                'project': self.project,
-                'version': self.version,
-                'master_doc': master_doc,
-            },
-        )
-        conf_file = os.path.join(docs_dir, 'conf.py')
-        safe_write(conf_file, conf_template)
```
Comment (stsewd) [inline readthedocs/doc_builder/backends/sphinx.py]: Looks like we can remove this function too

File: readthedocs/doc_builder/backends/sphinx.py [hunk 2]
```diff
@@ -168,14 +154,10 @@ def get_config_params(self):
...
     def append_conf(self, **__):
         """
-        Find or create a ``conf.py`` and appends default content.
+        Find a ``conf.py`` and appends default content.
 
         The default content is rendered from ``doc_builder/conf.py.tmpl``.
         """
-        if self.config_file is None:
-            master_doc = self.create_index(extension='rst')
-            self._write_config(master_doc=master_doc)
```
Comment (stsewd) [inline readthedocs/doc_builder/backends/sphinx.py]: We should raise an exception at this point https://github.com/rtfd/readthedocs.org/blob/43c1ba7e8b65bc3bb5a8b2d7b7f6de94dfa91b44/readthedocs/projects/models.py#L725-L725

Comment (saadmk11) [reply readthedocs/doc_builder/backends/sphinx.py]: @stsewd Do you mean we should keep this
https://github.com/rtfd/readthedocs.org/blob/30583646e470b1bdcf606c52e851835d6495df20/readthedocs/doc_builder/backends/sphinx.py#L176
then raise the error or just raise the error if `self.config_file is None`

**Comment by saadmk11** [comment]
@stsewd  Updated the PR according to your suggestions.

File: readthedocs/doc_builder/backends/sphinx.py (hunk 2 above)
Comment (stsewd) [reply readthedocs/doc_builder/backends/sphinx.py]: I feel that create index should stay, that's part of https://github.com/rtfd/readthedocs.org/issues/1800, we can address that in another PR

Comment (stsewd) [reply readthedocs/doc_builder/backends/sphinx.py]: (with or without config file)

File: readthedocs/rtd_tests/tests/test_doc_builder.py [hunk 3]
```diff
@@ -281,7 +254,7 @@ def test_get_theme_name_with_feature_flag(self, checkout_path, run):
 
     @patch('readthedocs.doc_builder.base.BaseBuilder.run')
     @patch('readthedocs.projects.models.Project.checkout_path')
-    def test_append_conf_create_yaml(self, checkout_path, run):
+    def test_append_conf_does_not_create_yaml(self, checkout_path, run):
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_doc_builder.py]: This should be named something like `test_mkdocs_yaml_not_found`

Comment (saadmk11) [reply readthedocs/rtd_tests/tests/test_doc_builder.py]: Okay, thanks :)

**Comment by saadmk11** [comment]
@stsewd Updated the PR

**Comment by saadmk11** [comment]
resolved conflicts

File: readthedocs/doc_builder/exceptions.py [hunk 4]
```diff
@@ -76,3 +76,7 @@ class MkDocsYAMLParseError(BuildEnvironmentError):
         'The "{config}" config from your MkDocs YAML config file has to be a '
         'a list of relative paths.',
     )
+    NOT_FOUND = ugettext_noop(
+        'A configuration file was not found. '
+        'Make sure you have a `mkdocs.yaml` file in your repository.',
```
Comment (stsewd) [inline readthedocs/doc_builder/exceptions.py]:
```suggestion
        'Make sure you have a "mkdocs.yaml" file in your repository.',
```

This is rendered as html, not markdown :)

File: readthedocs/rtd_tests/tests/test_doc_builder.py [hunk 5]
```diff
@@ -105,36 +98,17 @@ def test_create_conf_py(
             build_env=self.build_env,
             python_env=python_env,
         )
-        try:
+        with pytest.raises(ProjectConfigurationError):
             base_sphinx.append_conf()
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_doc_builder.py]: We should check that the msg error is the same as `ProjectConfigurationError.NOT_FOUND`

File: readthedocs/rtd_tests/tests/test_doc_builder.py [hunk 6]
```diff
@@ -295,40 +268,9 @@ def test_append_conf_create_yaml(self, checkout_path, run):
...
-        )
-        self.assertIsNone(
-            config['google_analytics'],
-        )
-        self.assertEqual(
-            config['site_name'],
-            'mkdocs',
-        )
+        with self.assertRaises(MkDocsYAMLParseError):
+            self.searchbuilder.append_conf()
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_doc_builder.py]: Same here, we should check the msg error.

File: readthedocs/doc_builder/exceptions.py (hunk 4 above)
Comment (saadmk11) [reply readthedocs/doc_builder/exceptions.py]: oh my bad :)

**Comment by saadmk11** [comment]
@stsewd Updated the PR :)

File: readthedocs/rtd_tests/tests/test_doc_builder.py [hunk 7]
```diff
@@ -295,40 +271,11 @@ def test_append_conf_create_yaml(self, checkout_path, run):
...
-        self.assertIsNone(
-            config['google_analytics'],
-        )
-        self.assertEqual(
-            config['site_name'],
-            'mkdocs',
-        )
+        with pytest.raises(
+                MkDocsYAMLParseError,
+                match=MkDocsYAMLParseError.NOT_FOUND
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_doc_builder.py]: Nice, didn't know about this option in pytest!

Comment (saadmk11) [reply readthedocs/rtd_tests/tests/test_doc_builder.py]: I learned it recently. :)

File: readthedocs/doc_builder/backends/sphinx.py [hunk 8]
```diff
@@ -177,6 +163,9 @@ def append_conf(self, **__):
         except IOError:
             raise ProjectConfigurationError(ProjectConfigurationError.NOT_FOUND)
 
+        # Create an index file if there is None.
+        self.create_index(extension='rst')
```
Comment (humitos) [inline readthedocs/doc_builder/backends/sphinx.py]: Considering that we are removing all these auto-generated config files, is this something that we still want?

If we are now requiring a `conf.py` in the project, it means that the user have already ran `sphinx-quickstart` which already generates a `index.rst`.

File: readthedocs/doc_builder/exceptions.py [hunk 9]
```diff
@@ -76,3 +76,7 @@ class MkDocsYAMLParseError(BuildEnvironmentError):
         'The "{config}" config from your MkDocs YAML config file has to be a '
         'a list of relative paths.',
     )
+    NOT_FOUND = ugettext_noop(
+        'A configuration file was not found. '
+        'Make sure you have a "mkdocs.yaml" file in your repository.',
```
Comment (humitos) [inline readthedocs/doc_builder/exceptions.py]: Not sure if it matters, but official MkDocs uses `mkdocs.yml` to refer this file. I suppose it's better to follow their convention here.

Comment (saadmk11) [reply readthedocs/doc_builder/exceptions.py]: @humitos Fixed. :)

File: readthedocs/doc_builder/backends/sphinx.py (hunk 8 above)
Comment (stsewd) [reply readthedocs/doc_builder/backends/sphinx.py]: This is another problem. Previously sphinx-quickstart used to generated another filename for index.

The issue to track this is https://github.com/rtfd/readthedocs.org/issues/1800

**Comment by stsewd** [comment]
> What is our plan for those projects?

@humitos It was decided that if those project don't build locally, they shouldn't build in rtd neither. See https://github.com/rtfd/readthedocs.org/issues/2483#issuecomment-484054492

**Comment by humitos** [comment]
@stsewd I think that is valid for new projects, but I don't think we want to break all the ones that are already building properly. Also, that seems like a comment from David to me, not a decision made.

Do we know how many projects that matches the conditions that I mentioned are currently building? I suppose that at least we should contact them and make a deprecation plan giving them some time to migrate.

**Comment by humitos** [comment]
We decided that,

1. we want to `log.info` the projects that are using this these _features_
1. communicate with them to help on the migration

Once we have collected this data and we gave some time to them to migrate, we can merge this PR. I'm marking as blocked until then.

**Comment by humitos** [comment]
We are ready to move forward with the plan I wrote in my previous comment https://github.com/readthedocs/readthedocs.org/pull/5609#issuecomment-497312034

So, the first step here is to add log lines to know what are the projects where we are creating a `conf.py` or a `mkdocs.yaml` file. It seems we already have this in place:

* Sphinx:
  * NR: https://onenr.io/0MR2yz7qbwY
  * Code: https://github.com/readthedocs/readthedocs.org/blob/37650de695dff4bd2e80fd957974959ca9379307/readthedocs/doc_builder/backends/sphinx.py#L118-L122
* MkDocs: 
  * NR: https://onenr.io/0oQDKbpp5jy
  * Code: https://github.com/readthedocs/readthedocs.org/blob/37650de695dff4bd2e80fd957974959ca9379307/readthedocs/doc_builder/backends/mkdocs.py#L121-L125

It seems we are ready to start contacting these users about the removal of this feature.

**Comment by ericholscher** [comment]
@humitos This needs to be merged after Aug 28 right?

**Comment by humitos** [comment]
Yes, https://blog.readthedocs.com/doctool-without-configuration-file/

**Comment by ericholscher** [comment]
@humitos Great, I've added it to the sprint for that timeframe 👍

File: readthedocs/doc_builder/backends/sphinx.py [hunk 10]
```diff
@@ -272,34 +249,29 @@ def get_config_params(self):
...
-                self.config_file or self.project.conf_file(self.version.slug)
-            )
+            if not os.path.exists(self.config_file):
+                raise UserFileNotFound
+
             # Allow symlinks, but only the ones that resolve inside the base directory.
             outfile = safe_open(
                 self.config_file, "a", allow_symlinks=True, base_path=self.project_path
             )
             if not outfile:
```
Comment (stsewd) [inline readthedocs/doc_builder/backends/sphinx.py:269]: the result from `safe_open` is always a context manager or an exception, so we can omit this check.

File: readthedocs/doc_builder/backends/sphinx.py [hunk 11]
```diff
@@ -272,34 +249,29 @@ def get_config_params(self):
...
-                )
-        except IOError as exc:
-            raise ProjectConfigurationError(
-                ProjectConfigurationError.NOT_FOUND
-            ) from exc
+                raise UserFileNotFound
+        except UserFileNotFound:
+            raise UserFileNotFound(
+                UserFileNotFound.FILE_NOT_FOUND.format(self.config_file)
+            )
```
Comment (stsewd) [inline readthedocs/doc_builder/backends/sphinx.py:274]: You can just raise the exception with the message below, and avoid the try..except block.

**Pull request merged**